caravel/views.py

```diff
@@ -1952,17 +1955,34 @@ def fetch_datasource_metadata(self):
             .filter_by(id=request.args.get('datasource_id'))
             .first()
         )
-        # SUPPORT DRUID
-        # TODO: move this logic to the model (maybe)
-        datasource_grains = datasource.database.grains()
-        grain_names = [str(grain.name) for grain in datasource_grains]
+
+        # Check if datasource exists
+        if not datasource:
+            return json_error_response(DATASOURCE_MISSING_ERR)
+        # Check permission for datasource
+        if not self.datasource_access(datasource):
+            return json_error_response(DATASOURCE_ACCESS_ERR)
+
+        time_columns = []
+        grains_choices = []
+        datasource_class_name = datasource_class.__name__
+        if datasource_class_name == 'SqlaTable':
+            time_columns = datasource.dttm_cols
+            grains = datasource.database.grains()
+            grains_choices = [grain.name for grain in grains]
+        elif datasource_class_name == 'DruidDatasource':
```

Comment:
please implement `@property` on the datasources in the models.py that will provide you the data you need in the uniform way for both druid and table